### ISSUE TITLE
Camera remains active after disconnection

### ISSUE THREAD

**Issue by andrewhl** [twilio/twilio-video.js#111]
We have a React Flux application with a `Preflight` component that is responsible for connecting participants to the video room. On `disconnected` and `participantDisconnected`, we disconnect from the room and detach the tracks. Unfortunately, the camera is staying active. Am I missing something?

Note: If one participant disconnects, in our implementation, we want all participants to disconnect. 

```js
    enterVideoRoom() {
        let videoToken = this.props.videoToken;
        const uuid = this.props.uuid;

        connect(videoToken.token, {
            name: uuid
        }).then(room => {
            VideoActionCreators.enterVideoRoom(room);

            window.onbeforeunload = () => {
                room.disconnect();
            };

            room.on('participantConnected', participant => {
                VideoActionCreators.beginVideoChat();
            });

            room.on('participantDisconnected', (participant) => {
                participant.tracks.forEach(track => {
                    let attachedElements = track.detach();
                    attachedElements.forEach(element => element.remove());
                });
                room.disconnect();
                VideoActionCreators.exitVideoRoom();
            });

            room.on('disconnected', room => {
                // Detach the local media elements
                room.localParticipant.tracks.forEach(track => {
                    let attachedElements = track.detach();
                    attachedElements.forEach(element => element.remove());
                });

                room.participants.forEach(participant => {
                    participant.tracks.forEach(track => {
                        let attachedElements = track.detach();
                        attachedElements.forEach(element => element.remove());
                    });
                });
                room.disconnect();
                VideoActionCreators.exitVideoRoom();
                VideoActionCreators.endVideoChat();
            });

        }, error => {
            console.log('Failed to connect to the room', error);
        });
    }
```

**Expected behavior:**

Camera deactivates for both participants.

**Actual behavior:**

Camera remains active for both participants.

**Software versions:**

 - [x] Browser(s): Chrome and Chrome Incognito Version 56.0.2924.87 (64-bit)
 - [x] Operating System: OSX 10.11.6
 - [x] twilio-video.js: 1.0.0
 - [x] Third-party libraries (e.g., Angular, React, etc.): React 15.5.4

**Comment by markandrus** [comment]
Hi @andrewhl,

When you call `connect` without the `tracks` option specified in [ConnectOptions](http://media.twiliocdn.com/sdk/js/video/releases/1.0.0/docs/global.html#ConnectOptions), `connect` behaves as follows:

> By default, this will automatically acquire an array containing a LocalAudioTrack and LocalVideoTrack before connecting to the Room. The LocalTracks will be stopped when you disconnect from the Room.

So in your code, when you call `connect`, if you were to pass `logLevel` equal to "debug",

```js
connect(videoToken.token, {
  name: uuid,
  logLevel: 'debug'
})
```

you should see lines like the following once you disconnect from the Room:

```
2017-05-10 19:58:08.308Z | INFO in [LocalAudioTrack #1: 3f9c18ad-7dc4-4221-8a9d-404c66fdad0d]: Stopping
log.js:152 2017-05-10 19:58:08.309Z | DEBUG in [LocalAudioTrack #1: 3f9c18ad-7dc4-4221-8a9d-404c66fdad0d]: Ended
2017-05-10 19:58:08.308Z | INFO in [LocalVideoTrack #1: 3f9c18ad-7dc4-4221-8a9d-404c66fdad0d]: Stopping
log.js:152 2017-05-10 19:58:08.309Z | DEBUG in [LocalVideoTrack #1: 3f9c18ad-7dc4-4221-8a9d-404c66fdad0d]: Ended
```

and if you were to check each of your Room's LocalParticipant's LocalTracks, you should see they are all stopped:

```js
room.localParticipant.tracks.forEach(track => console.log(track.isStopped));
// => true
// => true
```

Can you verify this behavior? If true, then I'd expect you might have a different LocalTrack somewhere in your application that is remaining active (e.g., for "previewing"?).

**Comment by andrewhl** [comment]
@markandrus Thanks again for your help! I was able to resolve the issue by: a) passing `localTracks` down from the `createLocalTracks()` success callback into my `connect()` function, via Flux state and props; and b) calling `track.stop()` as well as `track.detach()` when disconnecting.

One question: Is it safe to pass an error callback to `createLocalTracks().then()` as a second parameter? E.g.,

```js
        createLocalTracks().then(tracks => {
            ...
        }, error => {
            this.setState({videoError: true});
            Bugsnag.notify("Video Error", error);
        });
```

**Comment by markandrus** [comment]
@andrewhl great to hear! and

> One question: Is it safe to pass an error callback to createLocalTracks().then() as a second parameter?

Yes, it is--in fact it is best practice 👍 I'd recommend handling the error path in all of the asynchronous APIs (that includes `connect`, `createLocalTracks`, Room's `getStats`, etc.)
